spec.html

```diff
@@ -43009,7 +43009,7 @@ <h1>%AsyncFromSyncIteratorPrototype%.throw ( [ _value_ ] )</h1>
             1. If Type(_result_) is not Object, then
               1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; a newly created *TypeError* object &raquo;).
               1. Return _promiseCapability_.[[Promise]].
-            1. Return ! AsyncFromSyncIteratorContinuation(_result_, _promiseCapability_).
+            1. Return ! AsyncFromSyncIteratorContinuation(_result_, _promiseCapability_, _syncIteratorRecord_, *false*).
```

Comment:
Wouldn't you need to close here? Can you recover from a recovery?

Reply:
Good question. I find it hard to know what the "right" answer is for `throw`, since it doesn't come up as much (the user has to call it explicitly), but I think you're right. Compare, for example,

```js
function* count() {
  try {
    for (let i = 0; i < 10; ++i) {
      try {
        yield i;
      } catch (e) {
        console.log("caught and suppressed", e);
      }
    }
  } finally {
    console.log("time to clean up");
  }
}

function* wrap(inner) {
  yield* inner;
}

let iterable = wrap(count());
for (let item of iterable) {
  console.log(item);
  console.log("throw returned", iterable.throw("throwing here")); // note that this advances the iterator
  if (item >= 4) break;
}
```

This goes through three iterations of the loop, printing "caught and suppressed throwing here" in each one, and then on the last iteration, which `break`s, also prints "time to clean up".

(The call to `wrap` in this example has no observable effects, vs just `iterable = count()` which is my point - calling `.throw` on this wrapper delegates it to the inner iterator, which suppresses the exception, so the loop continues and can later call `.return`.)

---

I'm not at all clear on when you'd ever want to do call `throw` - I _think_ it was copied from Python, which uses specific kinds of exceptions for more general signals in a way that JS does not, but this was before my time and I haven't reviewed all the relevant notes. But if we assume the analogy above is reasonable, then I agree with @jridgewell.

Reply:
Good catch, I totally forgot about `throw` being able to recover. And I agree, letting the sync iterator decide is the right approach. Good thing is, since we now check the `done` value before installing the close on rejection logic, just switching the boolean to `true` here should be sufficient.

Which begs the question, should `return` be handled similarly? It seems that the result of `return` can similarly continue a `yield *` from my reading of the [steps](https://tc39.es/ecma262/#sec-generator-function-definitions-runtime-semantics-evaluation) (7.c.viii.)

Reply:
> Which begs the question, should `return` be handled similarly?

Huh. I guess so, looking at it, though I think this matters less - as [these slides](https://github.com/tc39/notes/blob/4592a41fea60cfec3148b26533d9b6df586b2a95/meetings/2014-06/closing-iterators.pdf) say, failure to stop iterating when `return` is called is probably a bug in the contract between the iterator and the loop.

(It would allow dropping the `closeIteratorOnRejection` parameter entirely, which is nice.)

Reply:
Based on
> - In short: any abrupt completion of the loop.
> - Normal completion should not call the method; in
that case the iterator itself decided to close

I'm willing to say, let's close if the sync iterator doesn't believe it's done, but the consumer of the wrapper considers the async iterator is buggy, regardless of what caused the iterator result to be yielded.

Reply:
I think I've argued myself back around to the other side of the `return` question - that is, I now think (weakly) we ought not close again when `.return` produces a rejected promise.

Right now, nothing in the language calls `throw` except to forward on explicit calls to `throw` from the user, so it's not obvious what guarantees anyone implementing the iterator protocol might expect or rely on about when `throw` gets called.

But `return` _does_ get called by the language. Specifically, `return` gets called exactly once whenever a language construct stops iterating a well-behaved iterator early, and the iterator is then discarded. (Except that it gets called zero times in the case addressed by this PR.) So that's the interface people are likely to be programming against.

Here we would be calling it twice. That seems like it would violate the expectation that it gets called exactly once. It's also not clear what benefit it would have: the language calls `return` to say "I'm done with this", but we've _already done that_ the first time we called `return`, and it's not like we're going to be any _more_ done just because the promise rejected.

Reply:
> except to forward on explicit calls to `throw` from the user

Yeah, and this is actually another inconsistency of this wrapper. If sync iterator doesn't have a `throw` method, the wrapper doesn't fallback to `.return` like `yield *` would.

> Here we would be calling it twice. That seems like it would violate the expectation that it gets called exactly once. It's also not clear what benefit it would have: the language calls `return` to say "I'm done with this", but we've _already done that_ the first time we called `return`, and it's not like we're going to be any _more_ done just because the promise rejected.

Yeah I think I agree.

@bakkot, what do you think of still calling `.return` if `.throw` returns a rejected promise as value with `done: false`, but not calling `.return` a second time if `.return` was already called, regardless if it returned `done: false` or not. In that case I can reintroduce the boolean flag and name it something like "returnAlreadyCalled".

Also, do you think we should add a `.return` fallback in the wrapper's `.throw`?

Reply:
I don't have much intuition for how `throw` is supposed to work, to be honest.

Reply:
> I don't have much intuition for how `throw` is supposed to work, to be honest.

Step `7.b.3` of `yield *` falls back to calling `.return` if `throw` is `undefined` on the iterator. The problem is that the wrapper has a `throw` on its prototype regardless of the shape of the sync iterator, so `yield *` would call the wrapper's `.throw`, which would simply reject because it can't find a `.throw` on the sync iterator, and not call `.return` like `yield *` would have if the wrapper lacked a `.throw`.

Reply:
Yeah, I guess the analogy to `yield*` is compelling. I'm on board with both 
- if `throw` is called on the outer iterator and the inner iterator lacks `throw`, fall back to `return`
- if `throw` is called on the outer iterator and the inner iterator has `throw` and it returns `{ done: false, value: Promise.reject() }`, call `return` on the inner iterator.

Reply:
@bakkot's suggestion SGTM.